tf/tfprocess.py

```diff
@@ -323,45 +284,90 @@ def replace_weights(self, new_weights):
             else:
+                # Can't populate renorm weights, but the current new_weight will need using elsewhere.
+                if 'renorm' in weights.name:
+                    offset-=1
+                    continue
+                # betas without gamms need to skip the gamma in the input.
+                if 'beta:' in weights.name and not last_was_gamma:
+                    source_idx+=1
+                    offset+=1
                 # Biases, batchnorm etc
-                new_weight = tf.constant(new_weights[e], shape=weights.shape)
-                all_evals.append(tf.assign(weights, new_weight))
-        self.session.run(all_evals)
+                new_weight = tf.constant(new_weights[source_idx], shape=weights.shape)
+                if 'stddev:' in weights.name:
+                    weights.assign(tf.math.sqrt(new_weight + 1e-5))
+                else:
+                    weights.assign(new_weight)
+                # need to use the variance to also populate the stddev for renorm, so adjust offset.
+                if 'variance:' in weights.name and self.renorm_enabled:
+                    offset-=1
+            last_was_gamma = 'gamma:' in weights.name
+        # Replace the SWA weights as well, ensuring swa accumulation is reset.
+        if self.swa_enabled:
+            self.swa_count.assign(tf.constant(0.))
+            self.update_swa_v2()
         # This should result in identical file to the starting one
-        # self.save_leelaz_weights('restored.txt')
+        # self.save_leelaz_weights_v2('restored.pb.gz')
 
-    def restore(self, file):
-        print("Restoring from {0}".format(file))
-        self.saver.restore(self.session, file)
+    def restore_v2(self):
+        if self.manager.latest_checkpoint is not None:
+            print("Restoring from {0}".format(self.manager.latest_checkpoint))
+            self.checkpoint.restore(self.manager.latest_checkpoint)
 
-    def process_loop(self, batch_size, test_batches, batch_splits=1):
+    def process_loop_v2(self, batch_size, test_batches, batch_splits=1):
         # Get the initial steps value in case this is a resume from a step count
         # which is not a multiple of total_steps.
-        steps = tf.train.global_step(self.session, self.global_step)
+        steps = self.global_step.read_value()
         total_steps = self.cfg['training']['total_steps']
         for _ in range(steps % total_steps, total_steps):
-            self.process(batch_size, test_batches, batch_splits=batch_splits)
+            self.process_v2(batch_size, test_batches, batch_splits=batch_splits)
+
+    @tf.function()
+    def read_weights(self):
+        return [w.read_value() for w in self.model.weights]
+
+    @tf.function()
+    def process_inner_loop(self):
+        print('tracing inner loop!')
+        x, y, z, q = next(self.train_iter)
+        with tf.GradientTape() as tape:
+            policy, value = self.model(x)
```

Comment:
`self.model` call should have `training=True` to set batch norm to training mode.

Reply:
Indeed - fixed.